### ISSUE TITLE
Similar saved_model for run_squad? 

### ISSUE THREAD

**Issue by apurvaasf** [bigboNed3/bert_serving#4]
Hi @bigboNed3 
Do you have similar files to export BERT for SQUAD? I am new to TensoFlow and hence having difficulty figuring out what the serving function for SQUAD should be. 
This will be a great help!

**Comment by jacky7788** [comment]
hi，you solved the problem？
I get the .pb now but idk how to make the prediction json file。。。only getting start end logit and the unique id

**Comment by apurvaasf** [comment]
Oh once we get start, end logit and unique ID, it's easy to get json. "write_predictions" function is the one that does the job. You can just copy what that does I think!
However, if I get how to get start and end logits, I can tell you exactly how to get the json. Can you tell me your steps to get start and end logits? Like the edit you made to the serving function and all?

**Comment by jacky7788** [comment]
after you export the model as .pb

CMD:
saved_model_cli show --all --dir $EXPORT_PATH

you will see 
![image](https://user-images.githubusercontent.com/28256291/51584826-fceff480-1f11-11e9-9013-2b271e8cd94a.png)



Not sure if it is correct...I am new to tensorflow too
the file I use to show start and end logits is:

`import tensorflow as tf
import numpy as np
from tensorflow.python.saved_model import tag_constants
import os
import collections
flags = tf.flags

FLAGS = flags.FLAGS

flags.DEFINE_integer(
    "max_seq_length", 128,
    "The maximum total input sequence length after WordPiece tokenization. "
    "Sequences longer than this will be truncated, and sequences shorter "
    "than this will be padded.")

flags.DEFINE_string(
    "export_dir", None,
    "The dir where the exported model has been written.")

flags.DEFINE_string(
    "predict_file", None,
    "The file of predict records")




graph = tf.Graph()
with graph.as_default():
    with tf.Session() as sess:
        tf.saved_model.loader.load(sess, [tag_constants.SERVING], FLAGS.export_dir)
        tensor_input_ids = graph.get_tensor_by_name('input_ids_1:0')
        tensor_input_mask = graph.get_tensor_by_name('input_mask_1:0')
        tensor_segment_ids = graph.get_tensor_by_name('segment_ids_1:0')
        tensor_unique_ids = graph.get_tensor_by_name('unique_ids_1:0')
        
        
        
        RawResult = collections.namedtuple("RawResult",
                               ["unique_id", "start_logits", "end_logits"])
        
        #tensor_outputs = graph.get_tensor_by_name('unique_ids_1:0')
        
        all_results=[]
        record_iterator = tf.python_io.tf_record_iterator(path=FLAGS.predict_file)
        for string_record in record_iterator:
            example = tf.train.Example()
            example.ParseFromString(string_record)
            input_ids = example.features.feature['input_ids'].int64_list.value
            input_mask = example.features.feature['input_mask'].int64_list.value
            segment_ids = example.features.feature['segment_ids'].int64_list.value
            unique_ids = example.features.feature['unique_ids'].int64_list.value
            
            tensor_outputs1 = graph.get_tensor_by_name('unstack:1')
            tensor_outputs2 = graph.get_tensor_by_name('unstack:0')
            tensor_outputs3 = graph.get_tensor_by_name('unique_ids_1:0')
            
            results1 = sess.run(tensor_outputs1, feed_dict={
                tensor_input_ids: np.array(input_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_input_mask: np.array(input_mask).reshape(-1, FLAGS.max_seq_length),
                tensor_segment_ids: np.array(segment_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_unique_ids: np.array(unique_ids),
            })
            results2 = sess.run(tensor_outputs2, feed_dict={
                tensor_input_ids: np.array(input_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_input_mask: np.array(input_mask).reshape(-1, FLAGS.max_seq_length),
                tensor_segment_ids: np.array(segment_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_unique_ids: np.array(unique_ids),
            })
            results3 = sess.run(tensor_outputs3, feed_dict={
                tensor_input_ids: np.array(input_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_input_mask: np.array(input_mask).reshape(-1, FLAGS.max_seq_length),
                tensor_segment_ids: np.array(segment_ids).reshape(-1, FLAGS.max_seq_length),
                tensor_unique_ids: np.array(unique_ids),
            })
    
            unique_id = int(results3)
            start_logits = [float(x) for x in results2.flat]
            end_logits = [float(x) for x in results1.flat]
            all_results.append(
                  RawResult(
                      unique_id=unique_id,
                      start_logits=start_logits,
                      end_logits=end_logits))
            
        print(all_results)
            
            
   
    
    `

**Comment by apurvaasf** [comment]
I'm working on this today. I will definitely let you know in a day or two. The solution with an explanation!

**Comment by jacky7788** [comment]
`  def serving_input_fn():
    input_ids = tf.placeholder(tf.int64, [None, FLAGS.max_seq_length], name='input_ids')
    input_mask = tf.placeholder(tf.int64, [None, FLAGS.max_seq_length], name='input_mask')
    segment_ids = tf.placeholder(tf.int64, [None, FLAGS.max_seq_length], name='segment_ids')
    unique_ids = tf.placeholder(tf.int64, [None, FLAGS.max_seq_length], name='unique_ids')
    
    
    unique_ids = tf.placeholder(tf.int64, [None], name='unique_ids')
    
    input_fn = tf.estimator.export.build_raw_serving_input_receiver_fn({
        'input_ids': input_ids,
        'input_mask': input_mask,
        'unique_ids': unique_ids,
        'segment_ids': segment_ids,
    })()
    return input_fn



  estimator._export_to_tpu = False
  estimator.export_savedmodel('.\\saved_model\\', serving_input_fn)`

**Comment by jacky7788** [comment]
hope this help =]

**Comment by apurvaasf** [comment]
And the predict_file that you give as input to test.py is? Is it a json?

**Comment by jacky7788** [comment]
eval.tf_record    and max_seq_length=384

**Comment by apurvaasf** [comment]
@jacky7788 If it is still unsolved, here you go:
Do you see the "write_predictions" function? That is all we need to call to get predictions.json from start and logits. Give it "all_results" that you generated and it will generate a "predictions.json" file.
Now the "write_predictions" function calls multiple other functions, which call other functions in turn. So if we keep the bare minimum number of functions to use the saved_model and then write the "predictions.json" file, the file becomes 950 lines big. 
Let me know if you face any issue with this!
